EIPS/eip-3540.md

```diff
@@ -0,0 +1,218 @@
+
+### Contract creation restrictions
+
+The [Changes to contact creation semantics](#changes-to-contract-creation-semantics) section defines minimal set of restrictions related to the contract creation: if _initcode_ or _code_ has the EOF1 container prefix it must be validated. This adds two validation steps in the contract creation, any of it failing will result in contract creation failure.
+
+Because _initcode_ and _code_ is evaluated for EOF1 independently, number of interesting combinations are allowed:
+- Create transaction with EOF1 _initcode_ can deploy legacy contract,
+- EOF1 contract can execute `CREATE` instruction with legacy _initcode_ to create new legacy contract,
+- Legacy contract can execute `CREATE` instruction with EOF1 _initcode_ to create new EOF1 contract,
+- Legacy contract can execute `CREATE` instruction with EOF1 _initcode_ to create new legacy contract,
+- etc.
+
+To limit the number of exotic bytecode version combinations, additional restrictions are considered, but current not part of the specification:
+
+1. The EOF version of _initcode_ must much the version of _code_.
+2. An EOF1 contract must not create legacy contracts.
+
+Finally, create transaction must be allowed to contain legacy _initcode_ and deploy legacy _code_ because otherwise there is no transition period allowing upgrading transaction signing tools. Deprecating such transactions may be considered in future.
+
+### The FORMAT byte
+
+The `0xEF` byte was chosen because it resembles **E**xecutable **F**ormat. It has a long history of being proposed for this use case, starting with [this](https://github.com/ethereum/EIPs/issues/154).
+
+### Section structure
+
+We have considered different questions for the sections:
+- Streaming headers (i.e. `section_header, section_data, section_header, section_data, ...`) are used in some other formats (such as WebAssembly). They are handy for formats which are subject to editing (adding/removing sections). That is not a useful feature for EVM. One minor benefit applicable to our case is they do not require a specific "header terminator".
+- Whether to have a header terminator or to encode `number_of_sections` or `total_size_of_headers`. Both of the latter raise the question how large of a value these fields should be able to hold. While today there will be only two sections, in case each "EVM function" would become their section, a fixed 8-bit field may not be big enough. A terminator byte seems to avoid these problems.
+- Whether to encode `section_size` as a fixed 16-bit value or some kind of variable length field (e.g. [LEB128](https://en.wikipedia.org/wiki/LEB128)). We have opted for fixed size, because it simplifies client implementations, and 16-bit seems enough, because of the currently exposed code size limit of 24576 bytes (see [EIP-170](./eip-170.md) and [EIP-2677](./eip-2677.md)). Should this be limiting in the future, a new EOF version could change the format.
```

Comment:
```suggestion
- Whether to encode `section_size` as a fixed 16-bit value or some kind of variable length field: We have opted for fixed size, because it simplifies client implementations, and 16-bit seems enough, because of the currently exposed code size limit of 24576 bytes (see [EIP-170](./eip-170.md) and [EIP-2677](./eip-2677.md)). Should this be limiting in the future, a new EOF version could change the format.
```

Also note: This creates a dependency on EIP-2677.  This may be fine, but consider removing the reference just to be safe.

Reply:
I'm unsure why rationale/motivation would create any dependency when linking to other EIPs. The only dependency on other EIPs happens when referral is done in the specification.

This has been the case for many EIPs so far and I don't see a good reason why gutting everything is a good approach. I appreciate that goal of slimming specifications, but I don't think in practice it has worked out to find all relevant sources of information if external links are not permitted. Kind of catch 22.